tools/chainsync-loadtest/src/concurrency/rate_limiter.rs

```diff
@@ -0,0 +1,54 @@
+use crate::concurrency::Ctx;
+use std::sync::Arc;
+use tokio::time;
+
+struct RateLimiter_ {
+    interval: time::Duration,
+    burst: u64,
+    tokens: u64,
+    ticks_processed: u64,
+    start: time::Instant,
+}
+
+impl RateLimiter_ {
+    fn ticks(&self, t: time::Instant) -> u64 {
+        return ((t - self.start).as_secs_f64() / self.interval.as_secs_f64()) as u64;
+    }
+    fn instant(&self, ticks: u64) -> time::Instant {
+        return self.start + self.interval.mul_f64(ticks as f64);
+    }
+}
+
+// RateLimiter is a Semaphore with periodically added permits.
+// It allows to rate limit any async-based operations.
+// It is parametrized by:
+// - interval - the amount of time after which a new permit is added.
+// - burst - the maximal number of permits in the semaphore.
+pub struct RateLimiter(Arc<tokio::sync::Mutex<RateLimiter_>>);
+
+impl RateLimiter {
+    pub fn new(interval: time::Duration, burst: u64) -> RateLimiter {
+        return RateLimiter(Arc::new(tokio::sync::Mutex::new(RateLimiter_ {
+            interval,
+            burst,
+            tokens: burst,
+            start: time::Instant::now(),
+            ticks_processed: 0,
+        })));
+    }
+
+    // See semantics of https://pkg.go.dev/golang.org/x/time/rate
+    pub async fn allow(&self, ctx: &Ctx) -> anyhow::Result<()> {
+        let mut rl = ctx.wrap(self.0.lock()).await?;
+        let ticks_now = rl.ticks(time::Instant::now());
+        rl.tokens = std::cmp::min(rl.burst, rl.tokens + (ticks_now - rl.ticks_processed));
```

Comment:
Wrapping arithmetic here when `ticks_now` approaches `u64::MAX`, I believe. Will `debug_assert!`. Use `wrapping_add/sub`.

Reply:
I want it to panic in such case. time::Instant also is capable of representing a limited time range and it is not like the time will stop afterwards. I'm not expecting for the code to work in such a case.